crates/runtime/src/lib.rs

```diff
@@ -568,6 +599,81 @@ impl Runtime {
             .register_metrics_table(self.prometheus_registry.is_some())
             .await?;
 
+        // Shutdown signal
+        let shutdown_signal_future = async {
+            let graceful_shutdown = async {
+                shutdown_signal().await;
+                tracing::debug!("Shutdown signal received. Press Ctrl-C again to force exit.");
+                self.shutdown().await;
+                Ok(())
+            };
+            tokio::select! {
+                result = graceful_shutdown => result,
+                () = force_shutdown_signal() => {
+                    tracing::info!("Force shutdown signal received. Terminating immediately.");
+                    // return error to force stop waiting for other tasks and terminate immediately
+                    Err(Error::ForceTerminated)
+                }
+            }
+        };
+
+        // - Scheduler: does some init, but all requests handled by Flight RPC stack
+        // - Executor: does some init, but has a polling loop to fetch work from scheduler
+        #[cfg(feature = "cluster")]
+        let maybe_cluster_future = match self.config.cluster.mode {
+            Some(ClusterMode::Scheduler) => {
+                cluster::initialize_cluster_scheduler(&self).await?;
+                None
+            }
+            Some(ClusterMode::Executor) => Some(
+                self.start_runtime_task(
+                    CLUSTER_EXECUTOR,
+                    None,
+                    cluster::initialize_cluster_executor(Arc::clone(&self)).await?,
+                )
+                .await,
+            ),
+            _ => None,
+        };
+
+        // Start Flight server
+        let flight_shutdown = CancellationToken::new();
+        let self_ref = Arc::clone(&self);
+        let cloned_tls_config = tls_config.clone();
+        let cloned_endpoint_auth = endpoint_auth.clone();
+        let cloned_app_ref = self_ref.app.read().await.as_ref().map(Arc::clone);
+
+        let flight_future = self
+            .start_runtime_task(FLIGHT_SERVER, Some(flight_shutdown.clone()), async move {
+                flight::start(
+                    config.flight_bind_address,
+                    cloned_app_ref,
+                    Arc::clone(&self_ref),
+                    cloned_tls_config,
+                    cloned_endpoint_auth,
+                    Arc::clone(&self_ref.rate_limits),
+                    Some(flight_shutdown),
+                )
+                .await
+                .context(UnableToStartFlightServerSnafu)
+            })
+            .await;
+
+        #[cfg(feature = "cluster")]
+        // If this is an executor, we only need the shutdown signal and flight server
+        if matches!(self.config.cluster.mode, Some(ClusterMode::Executor)) {
```

Comment:
Executors only run their work loop and the flight service. Graceful shutdown etc all work fine.